server/index.js

```diff
@@ -302,6 +302,110 @@ app.patch('/api/user/listings/:listingId', uploadsMiddleware, (req, res, next) =
     .catch(err => next(err));
 });
 
+app.post('/api/user/saved/:listingId', (req, res, next) => {
+  const listingId = parseInt(req.params.listingId, 10);
+  const { userId } = req.user;
+  if (!listingId) {
+    throw new ClientError(400, 'listingId must be a positive integer.');
+  }
+  const sql = `
+    insert into "savedItems"
+    ("listingId", "userId")
+    values ($1, $2)
+    returning *;
+    `;
+  const values = [listingId, userId];
+  db.query(sql, values)
+    .then(result => {
+      res.json(result.rows);
+    })
+    .catch(err => next(err));
+});
+
+app.get('/api/user/saved', (req, res, next) => {
+  const { userId } = req.user;
+  const sql = `
+    select *
+      from "listings"
+      join "savedItems" using ("listingId")
+      where "savedItems"."userId" = $1
+      order by "savedItems"."createdAt"
+    `;
+  const params = [userId];
+  db.query(sql, params)
+    .then(result => {
+      res.json(result.rows);
+    })
+    .catch(err => next(err));
+});
+
+app.get('/api/user/saved/listing/:listingId', (req, res, next) => {
+  const listingId = parseInt(req.params.listingId, 10);
+  const { userId } = req.user;
+  if (!listingId) {
+    throw new ClientError(400, 'listingId must be a positive integer.');
+  }
+  const sql = `
+    select *
+      from "listings"
+      join "savedItems" using ("listingId")
+      where "listings"."listingId" = $1
+        and "savedItems"."userId" = $2
+    `;
+  const params = [listingId, userId];
+  db.query(sql, params)
+    .then(result => {
+      if (!result.rows[0]) {
+        throw new ClientError(400, `Cannot find listing with listingId: ${listingId}`);
+      }
+      res.json(result.rows[0]);
+    })
+    .catch(err => next(err));
+});
+
+app.get('/api/user/saved/:listingId', (req, res, next) => {
+  const listingId = parseInt(req.params.listingId, 10);
+  const { userId } = req.user;
+  if (!listingId) {
+    throw new ClientError(400, 'listingId must be a positive integer.');
+  }
+  const sql = `
+    select *
+      from "savedItems"
+      where "listingId" = $1
+      and "userId" = $2
+    `;
+  const params = [listingId, userId];
+  db.query(sql, params)
+    .then(result => {
+      res.json(result.rows);
+    })
+    .catch(err => next(err));
+});
```

Comment:
What's the difference between these? Is there a specific reason to have both instead of just one of them?